### ISSUE TITLE
Miscellaneous bugs/fixes

### ISSUE THREAD

**Issue by rfortiz** [ccxt/ccxt#2465]
Thanks for creating this library, it's a real time saver. I've been using it with 3 exchanges so far and encountered a few issues (python3, ubuntu). Most are quite straight forward so I'll list them together, let me know if you want a separate ticket. Some solutions are for my personal usage and might not be suitable for all.

1. Bittrex: lot size is not defined, causing error when calling 'amount_to_lots'. --> suggest to set it equal to precision (1e-8) if lot is really not defined 
2. Bittrex: the balance structure is incomplete. entries 'indexed by currency first' do not exist if the market has not been traded before (instead of 'free', 'used', 'total' equal to zero)
3. Bittrex: 'fetch_my_trades' is not available and not emulated --> not sure if all infos are available but i'm emulating it using 'fetch_closed_orders' --> 'filled'
4. Cryptopia: 'fetch_my_trades' is limited to 100 by default, passing a higher limit in argument does not overide that. It is however supported by cryptopia API using additional 'params={'Count':99999}' --> I would expect this to be the default behavior, at least when a limit is passed in argument
5. Cryptopia: 1500 ms request limit is too high: https://www.cryptopia.co.nz/Forum/Thread/256 "1,000 requests/minute; 1,000,000 requests/day". --> got some nonce error if set to 60 ms, 120 ms seems fine although I'm far from using it that high on average. (I'm using  'nonce': ccxt.Exchange.milliseconds to instantiate the api)
6. Cryptopia: 'fetch_order' emulated, throws error when order is closed --> I find this behavior strange instead of returning 'status' equal to 'closed/canceled/unknown/not open' --> I ended up raising an exception for other exchanges as well every time the 'status' != 'open' so that I can later have the same code for all exchanges.
7. kucoin: 'fetch_my_trades' is limited to ~20 trades. the docs mentions a 'page' option. https://kucoinapidocs.docs.apiary.io/#reference/0/trading/list-dealt-orders(merged) --> maybe it's possible to get more, although it seems overly complicated for no reason.
8. kucoin:  'fetch_my_trades' returns trades without ids --> It seem available in the raw response: 'info' --> 'oid'

I hope this helps.

Finally there's one issue with cryptopia that I haven't been able to figure out. I'm calling 'fetch_tickers' in a timed loop which works fine most of the time, but once in a while get stuck with 'cryptopia fetchTickers() returned unrecognized pair id 5908'. The id is not always the same and I have to stop and restart my program.

**Comment by kroitor** [comment]
Hi!

Most of your questions are understandable, however, many of these issues are there by design due to different reasons:

> 1. Bittrex: lot size is not defined, causing error when calling 'amount_to_lots'. --> suggest to set it equal to precision (1e-8) if lot is really not defined

The `lot` size is not mentioned anywhere in the Manual, it is a deprecated field and it is to be removed, you should not rely on something that is not documented, really. We will use precision settings instead.

> 2. Bittrex: the balance structure is incomplete. entries 'indexed by currency first' do not exist if the market has not been traded before (instead of 'free', 'used', 'total' equal to zero)

Currencies that have not been funded at least once may be missing in the balance return, this will indicate if the currency has been traded at least once or not. If we force-set them to zero, we will lose partial info. Therefore you should check if it's there and add userland logic to treat it properly, as documented in the Manual: 

- https://github.com/ccxt/ccxt/wiki/Manual#balance-inference
- Also answered here and in similar questions on other exchanges: https://github.com/ccxt/ccxt/issues/2161

> 3. Bittrex: 'fetch_my_trades' is not available and not emulated --> not sure if all infos are available but i'm emulating it using 'fetch_closed_orders' --> 'filled'

We will not emulate fetchMyTrades with orders, because `order` != `trade`, as documented in the Manual: 

- https://github.com/ccxt/ccxt/wiki/Manual#personal-trades
- https://github.com/ccxt/ccxt/wiki/Manual#how-orders-are-related-to-trades

Bittrex will not return separate trades at all. Therefore you should use `fetchOrders`, `fetchOpenOrders`, `fetchClosedOrders` whatever is available. The same is true for all other exchanges. In fact, we are going to emulate as little as possible and probably will remove some buggy emulation in near future as well to make the library more robust.

> 4. Cryptopia: 'fetch_my_trades' is limited to 100 by default, passing a higher limit in argument does not overide that. It is however supported by cryptopia API using additional 'params={'Count':99999}' --> I would expect this to be the default behavior, at least when a limit is passed in argument

Fixed in ccxt 1.12.85, please update your version.

> 5. Cryptopia: 1500 ms request limit is too high: https://www.cryptopia.co.nz/Forum/Thread/256 "1,000 requests/minute; 1,000,000 requests/day". --> got some nonce error if set to 60 ms, 120 ms seems fine although I'm far from using it that high on average. (I'm using 'nonce': ccxt.Exchange.milliseconds to instantiate the api)

We are using a `safe default` there because Cryptopia has shown lags and packet drops earlier. The provided default rateLimit should be tweaked according to your needs and connection params... All about rate-limit: https://github.com/ccxt/ccxt/issues/1399

> 6. Cryptopia: 'fetch_order' emulated, throws error when order is closed --> I find this behavior strange instead of returning 'status' equal to 'closed/canceled/unknown/not open' --> I ended up raising an exception for other exchanges as well every time the 'status' != 'open' so that I can later have the same code for all exchanges.

This is documented here: 

- https://github.com/ccxt/ccxt/wiki/Manual#orders-cache
- https://github.com/ccxt/ccxt/wiki/Manual#error-handling

> 7. kucoin: 'fetch_my_trades' is limited to ~20 trades. the docs mentions a 'page' option. https://kucoinapidocs.docs.apiary.io/#reference/0/trading/list-dealt-orders(merged) --> maybe it's possible to get more, although it seems overly complicated for no reason.

You can use the kucoin pagination with their `page` extra param (just pass it in an assoc array as the last argument to the fetch_my_trades call). This is not in unified API for now.

> kucoin: 'fetch_my_trades' returns trades without ids --> It seem available in the raw response: 'info' --> 'oid'

Afaik, the `oid` is the order id, not a trade id. Therefore everything is correct, Kucoin trades don't have an id (at least kucoin does not send it), while their orders do have an id.

> Finally there's one issue with cryptopia that I haven't been able to figure out. I'm calling 'fetch_tickers' in a timed loop which works fine most of the time, but once in a while get stuck with 'cryptopia fetchTickers() returned unrecognized pair id 5908'. The id is not always the same and I have to stop and restart my program.

Will look into this issue and will report back here on my findings.

Hope this answers your questions. Feel free to reopen this issue if needed or just ask questions, if you have more. Thx!

**Comment by rfortiz** [comment]
Hi, thanks for the quick reply.

2. Missing keys seems like a very indirect way to find out if a pair has been traded at least once and is not consistent between exchanges. Anyway great to see you have plan to unify this.
6. I've read the doc on emulated 'fetch_order' and cache. I just find it strange that I have to raise an error for other exchanges to "unify" the behavior. Obviously it's only based on my use case and raising an error might be the best option in general context.
7. I tried to pass params={'page':2} to fetch_my_trades and got an empty array in return. I'm probably doing it wrong...
8. kucoin returns 2 ids: "orderOid" and "oid" --> Despite its name, I believe "oid' is unique and corresponds to what you call a trade or a fill

**Comment by kroitor** [comment]
> 6 - I've read the doc on emulated 'fetch_order' and cache. I just find it strange that I have to raise an error for other exchanges to "unify" the behavior. Obviously it's only based on my use case and raising an error might be the best option in general context.

Will look into it shortly.

7 - I tried to pass params={'page':2} to fetch_my_trades and got an empty array in return. I'm probably doing it wrong...

Can you show your verbose output of it? (without your keys)
https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-submit-an-issue

> 8 - kucoin returns 2 ids: "orderOid" and "oid" --> Despite its name, I believe "oid' is unique and corresponds to what you call a trade or a fill

Fixed in ccxt 1.12.86 )

**Comment by rfortiz** [comment]
7) I just got it, 'fetch_my_trades( symbol='AAA/BBB', params={'page':5})' requests page number 5 (and not 5 pages like I thought). If the page doesn't exist, ccxt only returns an empty array. To deal with it properly (without knowing the number of pages) one needs to check 'pageNos', 'currPageNo', (see verbose output). 

Such a pain, I don't understand why kucoin would complicate their API with pagination

```
import ccxt

api   = ccxt.kucoin({
    'apiKey': 'abcd',
    'secret': 'abcd',
    'verbose': True,
})

api_markets = api.load_markets()
trades = api.fetch_my_trades( symbol='AAA/BBB', params={'page':5})
print(trades)
```
verbose output when requesting pages that don't exist
```
Response: GET https://api.kucoin.com/v1/deal-orders?page=5&symbol=AAA-BBB 200 {'Date': 'Thu, 05 Apr 2018 14:56:02 GMT', 'Content-Type': 'application/json;charset=UTF-8', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Access-Control-Allow-Origin': 'https://www.kucoin.com', 'Access-Control-Allow-Credentials': 'true', 'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept, Accept-Language', 'Access-Control-Allow-Methods': 'POST, GET, PUT, OPTIONS, DELETE, PATCH', 'Access-Control-Max-Age': '86400', 'KC-NODE': 'localhost'} {"success":true,"code":"OK","msg":"Operation succeeded.","timestamp":1522940162321,"data":{"total":48,"firstPage":false,"lastPage":true,"datas":[],"currPageNo":4,"limit":12,"pageNos":4}}
[]
```
